csharp/src/Ice/LocatorInfo.cs

```diff
@@ -90,83 +90,81 @@ internal void ClearCache(Reference reference)
         }
 
         /// <summary>Resolves an indirect reference using the locator proxy or cache.</summary>
-        internal async ValueTask<(EndpointList Endpoints, bool Cached)> ResolveIndirectReferenceAsync(
+        internal async ValueTask<(EndpointList Endpoints, TimeSpan EndpointsAge)> ResolveIndirectReferenceAsync(
             Reference reference,
+            TimeSpan endpointsMaxAge,
             CancellationToken cancel)
         {
             Debug.Assert(reference.IsIndirect);
 
             EndpointList endpoints = ImmutableArray<Endpoint>.Empty;
-            bool cached = false;
+            TimeSpan endpointsAge = TimeSpan.Zero;
 
             Location location = reference.Location;
-
             if (reference.IsWellKnown)
             {
                 // First, we check the cache.
-                (endpoints, location, cached) = GetResolvedWellKnownProxyFromCache(reference,
-                                                                                   reference.LocatorCacheTimeout);
-                if (!cached)
+                (endpoints, location, endpointsAge) = GetResolvedWellKnownProxyFromCache(reference);
+                bool expired = CheckExpired(endpointsAge, reference.LocatorCacheTimeout);
+                // If no endpoints are returned from the cache, or if the cache returned and expired endpoint and
+                // background updates are disabled, or if the caller is requesting a more recent endpoint that the
+                // one returned from the cache, we try to resolve the endpoint again.
+                if ((endpoints.Count == 0 && location.Count == 0) ||
+                    (!_background && expired) ||
+                    (endpointsMaxAge != Timeout.InfiniteTimeSpan && endpointsAge >= endpointsMaxAge))
```

Comment:
The first call pass `endpointsMaxAge = Timeout.InfiniteTimeSpan` and if we don't have this check it will query the locator even if the cached endpoint is still valid

Reply:
Ok, `Timeout.InfiniteTimeSpan` is actually defined as `-1ms` so this is indeed necessary. Rather than calling this method with an infinite timeout, we could pass `TimeSpan.MaxValue` to indicate that we always want to use the cached value and we could get rid of this explicit check on `Timeout.InfiniteTimeSpan`.

Reply:
I was thinking it would be simpler to pass the Min value of LocatorCacheTimeout or the last resolved time from the caller, the difficulty is with background updates as we allow to use an expired endpoint and queue the update to run in the background.

Reply:
Note we also use -1 InfiniteTimeSpan for the cache timeout, that is the main reason I pick that, it seems a bit more consistent.